examples/eventsub_websocket/src/websocket.rs

```diff
@@ -1,170 +1,359 @@
+                        match side_effect {
+                            Action::Nothing => {}
+                            Action::ResetKeepalive => timeout = Instant::now() + Duration::from_secs(WINDOW),
+                            Action::KillPredecessor => kill_predecessor_tx.send(())?,
+                            Action::AssignSuccessor(actor_handle) => {
+                                successor = Some(actor_handle);
+                            },
                         }
-                        Ok(())
-                    }
-                    EventsubWebsocketData::Revocation {
-                        metadata,
-                        payload: _,
-                    } => eyre::bail!("got revocation event: {metadata:?}"),
-                    EventsubWebsocketData::Keepalive {
-                        metadata: _,
-                        payload: _,
-                    } => Ok(()),
-                    _ => Ok(()),
+                    },
+                    _ = tokio::time::sleep_until(timeout) => eyre::bail!("connection timed out"),
                 }
             }
-            tungstenite::Message::Close(_) => todo!(),
-            _ => Ok(()),
-        }
+        }))
     }
 
-    pub async fn process_welcome_message(
-        &mut self,
-        data: SessionData<'_>,
-    ) -> Result<(), eyre::Report> {
-        self.session_id = Some(data.id.to_string());
-        if let Some(url) = data.reconnect_url {
-            self.connect_url = url.parse()?;
-        }
-        // check if the token is expired, if it is, request a new token. This only works if using a oauth service for getting a token
-        if self.token.is_elapsed() {
-            self.token =
-                crate::util::get_access_token(self.client.get_client(), &self.opts).await?;
+    pub async fn join(self) -> Result<eyre::Result<Self>, JoinError> { self.0.await }
+}
+
+pub async fn run(
+    helix_client: &'static HelixClient<'_, Client>,
+    token: UserToken,
+    opts: Arc<crate::opts::Opts>,
+    user_id: types::UserId,
+) -> eyre::Result<()> {
+    let url = twitch_api::TWITCH_EVENTSUB_WEBSOCKET_URL.clone();
+    let token = Arc::new(Mutex::new(token));
+    let user_id = Arc::new(user_id);
+    let subscribed = Arc::new(AtomicBool::new(false));
+
+    /*
+    when the active actor receives a reconnect message
+    it is supposed to spawn a new connection
+    and wait for it to receive a welcome message before closing.
+    we achieve this by giving a newly spawned actor a sender to kill the current connection.
+
+    https://dev.twitch.tv/docs/eventsub/handling-websocket-events/#reconnect-message
+
+    since this is a root actor without a predecessor it has no previous connection to kill
+    but we still need to give it a sender to satisfy the function signature.
+    `_` and `_unused` have different semantics where `_` is dropped immediately and sender gets a recv error
+     */
```

Comment:
try to not use /* when paragraphing, see https://doc.rust-lang.org/style-guide/index.html#comments

I also think this could be rewritten and should be moved to somewhere else, as what it describes is not only this specific call but every new invocations of ActorHandle::spawn

```
// If we receive a reconnect message we want to spawn a new connection to twitch.
// The already existing session should wait on the new session to receive a welcome message before being closed.
```